### PR TITLE
Issue297 continued

### LINKED ISSUE #297: More algorithms for Generic Acoustic Components
I will finish off the Stacked Harmonics algorithm and create 2 or 3 new ones, e.g. a positive-sloped  ridge and a negative-sloped ridge algorithm and a spectral-peak-track algorithm. The horizontal ridge algorithm would be pretty much equivalent to the existing whistle recognizer. There are possibly other component recognizers worth trying.

I will create a new branch in which all of these will be included.

### PR DESCRIPTION
**Pull request by towsey**
# Title of PR
Merge this branch into MAster.

What is the purpose of this PR?
Want to merge with master so I can get on with writing new recognizers.

Fixes #297 

## Changes
Fixed remaining bugs with Event merging and drawing o event labels and score bars.

## Issues
Do not believe so.

## Visual Changes

No
## Final Checklist

-   [ ] Assign reviewers if you have permission
-   [ ] Assign labels if you have permission
-   [x] Link issues related to PR
-   [ ] Link any PRs or issues blocking this PR from being merged
-   [ ] Remove/Reduce warnings from edited files
-   [ ] Unit tests have been added for changes
-   [ ] Ensure CI build is passing

### CODE REVIEW

File: src/AudioAnalysisTools/Events/Drawing/EventDrawer.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            {
+                return;
+            }
+
+            // TODO: add a Interval<double> ScoreRange property to EventCommon
+            // so we can properly normalize this value to the unit value.
+            // For now, we just assume it is normalized to [0,1].
+            //var clampedScore = @event.Score.Clamp(0, 1);
+            var normalisedScore = @event.ScoreNormalised;
+
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Drawing/EventDrawer.cs:36]: Remove comment - this has been no?

Also, the value should be clamped unless ScoreNormalised guarantees a value of [0,1]

File: src/AudioAnalysisTools/Events/Drawing/EventDrawer.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                return;
+            }
+
+            var rect = options.Converters.GetPixelRectangle(@event);
+
+            var scaledHeight = (float)normalisedScore * rect.Height;
+
+            graphics.NoAA().DrawLines(
+                options.Score,
+                new PointF(rect.Left, rect.Bottom - 1), // TODO minus one is a hack! just to make it work!
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Drawing/EventDrawer.cs:48]: Remove hack. Document why there's `- 1` there

File: src/AudioAnalysisTools/Events/EventCommon.cs [hunk 3]
```diff
@@ -4,13 +4,51 @@
...
+        public override double Score { get; set; }
+
+        /// <summary>
+        /// Gets or sets a min-max range of values for the score for this event.
+        /// This is used to establish a score scale and thereby normalise the score.
+        /// By default the minimum value of range = zero.
+        /// </summary>
+        public Interval<double> ScoreRange { get; set; }
+
+        public double ScoreNormalised => this.Score / this.ScoreRange.Maximum;
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/EventCommon.cs:50]: Incorrect calculation. Use:
```suggestion
        public double ScoreNormalised => ((this.Score - this.ScoreRange.Minimum) / this.ScoreRange.Size());
```

File: src/AudioAnalysisTools/Events/Tracks/PointData.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            // assumes a SortedSet is used (and that iteration order is signficant, unlike with HashSet)
+            var path = this.Points.Select(x => options.Converters.GetPoint(x)).ToArray();
+
+            // note not using AA here
+            // note could base pen thickness off ISpectralPoint thickness for a more accurate representation
+            graphics.DrawLines(
+                options.Border,
+                path);
+        }
+    }
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Tracks/PointData.cs:42]: Duplicate code and method from `IPointData` - remove this file

File: src/AudioAnalysisTools/Events/Types/ChirpEvent.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        /// MaxScore establishes a scale for the chirp score. Typically the amplitude of track points is decibels.
+        /// A satisfactory maxScore is 12.0 decibels, since this is a high SNR in enviornmental recordings.
+        /// The normalised score is a linear conversion from 0 - maxScore to [0, 1].
+        /// </remarks>
+        /// <param name="chirp">A chirp track consisting of a sequence of spectral points.</param>
+        /// <param name="maxScore">A maximum score used to normalise the track score.</param>
+        public ChirpEvent(Track chirp, double maxScore)
+        {
+            this.Tracks.Add(chirp);
+            this.ScoreRange = new Interval<double>(0, maxScore);
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Types/ChirpEvent.cs:34]: `maxScore` should be an interval.

Note you can use any of the following syntaxes where you need to create an interval:

- `(0, someNumber)`
- `someNumber.AsIntervalFromZero()`
- `0.AsIntervalTo(someNumber)`
- `new Interval<double>(0, someNumber)`

File: src/AudioAnalysisTools/Events/Types/ChirpEvent.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        /// Gets the normalised value for the event's track score.
+        /// NOTE: It is assumed that the minimum value of the score range = zero.
+        /// </summary>
+        public double ScoreNormalised
+        {
+            get
+            {
+                return this.Score / this.maxScore;
+            }
+        }
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Types/ChirpEvent.cs:75]: this looks to be redundant?

File: src/AudioAnalysisTools/Events/Types/ChirpEvent.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+        public override void Draw(IImageProcessingContext graphics, EventRenderingOptions options)
+        {
+            this.Tracks.First().Draw(graphics, options);
+
+            //  base drawing (border)
+            // TODO: unless border is disabled
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Types/ChirpEvent.cs:82]: remove TODO, we've done that

File: src/AudioAnalysisTools/Events/Types/ClickEvent.cs [hunk 8]
```diff
@@ -4,20 +4,63 @@
...
+                return this.Tracks.Single().GetAverageTrackAmplitude();
+            }
+        }
 
         public override void Draw(IImageProcessingContext graphics, EventRenderingOptions options)
         {
-            // TODO: render click event
+            // foreach (var track in tracks) {
+            // track.Draw(...)
+            // }
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Types/ClickEvent.cs:57]: Remove
```suggestion

```

File: src/AudioAnalysisTools/Events/Types/ClickEvent.cs [hunk 9]
```diff
@@ -4,20 +4,63 @@
...
 
+            this.Tracks.First().Draw(graphics, options);
 
-            // don't call base draw method because don't want the border
+            //  base drawing (border)
+            // TODO: unless border is disabled
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Types/ClickEvent.cs:62]:
```suggestion
```

File: src/AudioAnalysisTools/Events/Types/CompositeEvent.cs [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+                    {
+                        foreach (var track in eventWithTracks.Tracks)
+                        {
+                            yield return track;
+                        }
+                    }
+                }
+            }
+        }
+
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Types/CompositeEvent.cs:59]: `ScoreNormalised` probably needs to be overridden here to be the product of the component `ScoreNormalized`

File: src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs [hunk 11]
```diff
@@ -23,20 +23,20 @@ namespace AudioAnalysisTools.LongDurationSpectrograms
     /// It was written to deal with a set of recordings with protocol of Gianna Pavan (10 minutes every 30 minutes).
     ///
     /// The following Powershell command was constructed by Anthony to do the analysis and join the sequence of images so derived:
-    /// Y:\Italy_GianniPavan\Sassofratino1day | % {& "C:\Work\GitHub\audio-analysis\AudioAnalysis\AnalysisPrograms\bin\Release\AnalysisPrograms.exe" audio2csv -so ($_.FullName) -o "Y:\Italy_GianniPavan\output" -c "C:\Work\GitHub\audio-analysis\AudioAnalysis\AnalysisConfigFiles\Towsey.Acoustic.Parallel.yml" }
+    /// Y:\Italy_GianniPavan\Sassofratino1day | % {"&" "C:\Work\GitHub\audio-analysis\AudioAnalysis\AnalysisPrograms\bin\Release\AnalysisPrograms.exe" audio2csv -so ($_.FullName) -o "Y:\Italy_GianniPavan\output" -c "C:\Work\GitHub\audio-analysis\AudioAnalysis\AnalysisConfigFiles\Towsey.Acoustic.Parallel.yml" }
```
Comment (atruskie) [inline src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs:26]:
```suggestion
    /// Y:\Italy_GianniPavan\Sassofratino1day | % { &amp; "C:\Work\GitHub\audio-analysis\AudioAnalysis\AnalysisPrograms\bin\Release\AnalysisPrograms.exe" audio2csv -so ($_.FullName) -o "Y:\Italy_GianniPavan\output" -c "C:\Work\GitHub\audio-analysis\AudioAnalysis\AnalysisConfigFiles\Towsey.Acoustic.Parallel.yml" }
```

File: src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs [hunk 12]
```diff
@@ -23,20 +23,20 @@ namespace AudioAnalysisTools.LongDurationSpectrograms
...
     /// where:
     ///         Y:\Italy_GianniPavan\Sassofratino1day   is the directory containing recordings
     ///         | = a pipe
     ///         % = foreach{}  = perform the operation in curly brackets on each item piped from the directory.
-    ///         & "C:\Work\GitHub\audio-analysis\AudioAnalysis\AnalysisPrograms\bin\Release\AnalysisPrograms.exe"  = runs an executable
+    ///         "&" "C:\Work\GitHub\audio-analysis\AudioAnalysis\AnalysisPrograms\bin\Release\AnalysisPrograms.exe"  = runs an executable
```
Comment (atruskie) [inline src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs:31]:
```suggestion
    ///         &amp; "C:\Work\GitHub\audio-analysis\AudioAnalysis\AnalysisPrograms\bin\Release\AnalysisPrograms.exe"  = runs an executable
```

File: src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs [hunk 13]
```diff
@@ -23,20 +23,20 @@ namespace AudioAnalysisTools.LongDurationSpectrograms
...
     ///         audio2csv = first command line argument which determines the "activity" performed
     ///         -so ($_.FullName)  = the input file
     ///         -o "Y:\Italy_GianniPavan\output" = the output directory
     ///         -c "PATH\Towsey.Acoustic.Parallel.yml" is the config file
     ///
     /// The following PowerShell command was used by Anthony to stitch together a sequence of spectrogam images without any gap between them.
     /// It requires ImageMagick software to be installed: i.e. C:\Program Files\ImageMagick-6.8.9-Q16\montage.exe
-    /// Y:\Italy_GianniPavan\output\Towsey.Acoustic> & "C:\Program Files\ImageMagick-6.8.9-Q16\montage.exe" -mode concatenate -tile x1 *2MAP* "..\..\merge.png"
+    /// Y:\Italy_GianniPavan\output\Towsey.Acoustic> "&" "C:\Program Files\ImageMagick-6.8.9-Q16\montage.exe" -mode concatenate -tile x1 *2MAP* "..\..\merge.png"
```
Comment (atruskie) [inline src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs:39]:
```suggestion
    /// Y:\Italy_GianniPavan\output\Towsey.Acoustic> &amp; "C:\Program Files\ImageMagick-6.8.9-Q16\montage.exe" -mode concatenate -tile x1 *2MAP* "..\..\merge.png"
```

File: src/AudioAnalysisTools/Scales/LinearSecondsScale.cs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+        public LinearSecondsScale(int sampleRate, int frameSize, int stepSize)
+        {
+            this.SecondsPerFrame = frameSize / (double)sampleRate;
+            this.SecondsPerFrameStep = stepSize / (double)sampleRate;
+        }
+
+        public double SecondsPerFrame { get; }
+
+        public double SecondsPerFrameStep { get; }
+    }
```
Comment (atruskie) [inline src/AudioAnalysisTools/Scales/LinearSecondsScale.cs:24]: This class is unused, remove

File: src/AudioAnalysisTools/Tracks/ForwardTrackAlgorithm.cs [hunk 15]
```diff
@@ -217,6 +227,9 @@ public static SpectralTrack GetTrack(double[,] peaks, int startRow, int startBin
                 }
 
                 track.SetPoint(row, bin, maxValue);
+
+                // next line is for debug purposes
+                var info = track.CheckPoint(row, bin);
```
Comment (atruskie) [inline src/AudioAnalysisTools/Tracks/ForwardTrackAlgorithm.cs:232]: Remove or disable debug code before master

File: tests/Acoustics.Test/AnalysisPrograms/Recognizers/PteropusSp/PteropusSpTests.cs [hunk 16]
```diff
@@ -104,7 +106,7 @@ public void TestGetWingBeatEvents()
             Assert.AreEqual(46, acousticEvents[3].Oblong.ColumnRight);
             Assert.AreEqual(2334, acousticEvents[3].Oblong.RowTop);
             Assert.AreEqual(2383, acousticEvents[3].Oblong.RowBottom);
-
+            */
             //Assert.AreEqual(0.6062, stats.SpectralEnergyDistribution, 1E-4);
```
Comment (atruskie) [inline tests/Acoustics.Test/AnalysisPrograms/Recognizers/PteropusSp/PteropusSpTests.cs:110]: Needs to be fixed

File: src/AudioAnalysisTools/Events/Drawing/EventDrawer.cs (hunk 1 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/Drawing/EventDrawer.cs:36]: This issue has been fixed in latest commit.
Comment has been changed and clamp shifted to the Draw method.

File: src/AudioAnalysisTools/Events/Drawing/EventDrawer.cs (hunk 2 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/Drawing/EventDrawer.cs:48]: Comment has been added.

File: src/AudioAnalysisTools/Events/EventCommon.cs (hunk 3 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/EventCommon.cs:50]: This matter has been dealt with in the above changes. Typically the ScoreRange.Minimum will = zero Therefore subtracting zero will not make a difference.

File: src/AudioAnalysisTools/Events/Tracks/PointData.cs (hunk 4 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/Tracks/PointData.cs:42]: FIle has been removed and code still builds!

File: src/AudioAnalysisTools/Events/Types/ChirpEvent.cs (hunk 5 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/Types/ChirpEvent.cs:34]: This constructor has been changed for all four track type events.

File: src/AudioAnalysisTools/Events/Types/ChirpEvent.cs (hunk 6 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/Types/ChirpEvent.cs:75]: Has been removed.

File: src/AudioAnalysisTools/Events/Types/ChirpEvent.cs (hunk 7 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/Types/ChirpEvent.cs:82]: Have removed

File: src/AudioAnalysisTools/Events/Types/ClickEvent.cs (hunk 8 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/Types/ClickEvent.cs:57]: comments removed.

File: src/AudioAnalysisTools/Events/Types/ClickEvent.cs (hunk 9 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/Types/ClickEvent.cs:62]: Removed.

File: src/AudioAnalysisTools/Events/Types/CompositeEvent.cs (hunk 10 above)
Comment (towsey) [reply src/AudioAnalysisTools/Events/Types/CompositeEvent.cs:59]: ScoreNormalised has been overridden to return normalised average of component scores.

File: src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs (hunk 11 above)
Comment (towsey) [reply src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs:26]: This and other changes made to this file.

File: src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs (hunk 12 above)
Comment (towsey) [reply src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs:31]: done

File: src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs (hunk 13 above)
Comment (towsey) [reply src/AudioAnalysisTools/LongDurationSpectrograms/LDSpectrogramClusters.cs:39]: done

**Comment by atruskie** [comment]
The changes requested in this PR have been made by @towsey in https://github.com/QutEcoacoustics/audio-analysis/tree/BoobookRecognizerEtc. This PR is closed in favour for that branch

**Pull request merged**